### PR TITLE
OF-2655: Fix S2S outbound hanging

### PR DESCRIPTION
**Pull request by AlexGidman**
- Explicitly add close listener for new sessions that are created.
- Fix the new session on stream ID code for outbound S2S by verifying the streamID is new.
- Clear routing table cache of any lingering session objects via a replaceServerRoute function in the RoutingTable for atomic session replacement

### CODE REVIEW

**Comment by guusdk** [comment]
If the code is creating a second `Session` instance, shouldn't we at some point need a bit of code that removes references to the old session from things like the Session manager?

**Comment by AlexGidman** [comment]
> that removes references to the old session from things like the Session manager?

I may be incorrect, but I was struggling to find code that already does this; it appeared that we only remove server routes from the routing table, and do not cache outgoing server sessions anywhere?

**Comment by guusdk** [comment]
For posterity: Alex and me had an chat. This data appears to be stored in the routing table. The approach that we're looking at is replacing the routing table entry. First by removing and adding an entry, to validate the approach. If that works, an atomic 'replace' method should be added to the RoutingTable's API.

File: xmppserver/src/main/java/org/jivesoftware/openfire/spi/RoutingTableImpl.java [hunk 1]
```diff
@@ -227,18 +227,38 @@ public void addServerRoute(DomainPair address, LocalOutgoingServerSession destin
...
-            }
+            checkOnlyOneClusterNodeConnectedToAddress(address);
         }
         finally {
             lock.unlock();
         }
         localRoutingTable.addRoute(address, destination);
     }
 
+    private void checkOnlyOneClusterNodeConnectedToAddress(DomainPair address) {
```
[the remark this answers is not in the document]
Comment (guusdk) [reply xmppserver/src/main/java/org/jivesoftware/openfire/spi/RoutingTableImpl.java:238]: Also - doesn't this leave the cache empty (or at least: without an entry for domainPair)?

All of the other code is modifying the localRoutingTable outside of the lock. Let's be consistent and replicate that behavior.

[the remark this answers is not in the document]
Comment (guusdk) [reply xmppserver/src/main/java/org/jivesoftware/openfire/spi/RoutingTableImpl.java:238]: Interestingly, when I refactor the new `replaceServerRoute` method based on the feedback above, I'm getting an implementation that is an exact duplicate of the pre-existing `addServerRoute`!

`addServerRoute`:
- throws an exception when another cluster node already recorded the cache entry
- adds the local cluster identifier to the cache if that's not the case.
- replaces the value in the `local` table otherewise.

We might get away with just using the original `add` method. If we do, we should at least update its javadoc to make sure that the 'add-or-update' behavior that we'll depend on is part of the API contract.

[the remark this answers is not in the document]
Comment (guusdk) [reply xmppserver/src/main/java/org/jivesoftware/openfire/spi/RoutingTableImpl.java]: I've just pushed a commit that implements the above - basically it removes the newly added API again.

**Pull request merged**